src/axom/primal/geometry/KnotVector.hpp

```diff
@@ -0,0 +1,702 @@
+// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file KnotVector.hpp
+ *
+ * \brief A class to represent knot vectors for NURBS
+ */
+
+#ifndef AXOM_PRIMAL_KNOTVECTOR_HPP
+#define AXOM_PRIMAL_KNOTVECTOR_HPP
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+
+#include <ostream>
+#include "axom/fmt.hpp"
+
+namespace axom
+{
+namespace primal
+{
+// Forward declare the templated classes and operator functions
+template <typename T>
+class KnotVector;
+
+/*! \brief Overloaded output operator for knot vectors */
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const KnotVector<T>& kVec);
+
+/*!
+ * \class KnotVector
+ *
+ * \brief Represents the knot vector for a B-Spline/NURBS Curve/Surface
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * 
+ * Contains methods for knot vector manipulation shared between NURBS curves and surfaces.
+ * This class maintains the degree and requirements on a knot vector, namely
+ *  > Monotonicity (u_i <= u_i+1)
+ *  > Open/Clamped-ness (degree + 1 knots at beginning and end)
+ *  > Continuity (maximum internal knot multiplicity = degree)
+ */
+template <typename T>
+class KnotVector
+{
+public:
+  AXOM_STATIC_ASSERT_MSG(
+    std::is_arithmetic<T>::value,
+    "A knot vector must be defined using an arithmetic type");
+
+public:
+  /// \brief Default constructor
+  KnotVector() : m_deg(-1) { m_knots.resize(0); }
```

Comment:
Might be worth mentioning that the default-constructed `KnotVector `is not valid.

Reply:
Good idea, I've clarified this behavior a little bit by making a separate default constructor that always makes an "empty" curve, as distinct from a "degree -1" curve.